Rishabh/clarify questions system prompt

rishsriv opened the pull request:
- made phrasing clearer for 3 questions where we wanted all months/years - including those which did not occur in the data. This does not change results for any of the models, but figured we might as well clean it up
- updated the golden query for a question (`Return the highest sale price for each make and model of cars that have been sold and are no longer in inventory, ordered by the sale price from highest to lowest.`)
- added a system prompt to `prompt_cot.md`
- fixed anthropic and bedrock runners
- fixed an answer for an advanced question where it was unnecessarily coalescing a value to 0

rishsriv commented:
JP/Wendy - for whenever you review this, feel free to make any changes needed and merge!

wongjingping commented on data/instruct_advanced_postgres.csv:
```diff
@@ -41,7 +41,7 @@ SPM (Selling Profit Margin) = (Total Amount from Sells - (Tax + Commission)) / T
 CR = Rank customers by their total transaction volume, identifying the customer with the highest transaction volume as rank 1. This involves joining price data with ticker identifiers and filtering for a specified date range."
 car_dealership,instructions_cte_join,"What is the average number of days between the sale date and payment received date, rounded to 2 decimal places?","When getting duration between sale and payment date for each sale, get the latest payment for sale by aggregating over the payments_received table first.","WITH sale_payments AS (SELECT s.id AS sale_id, s.sale_date, MAX(p.payment_date) AS latest_ ...
-car_dealership,instructions_cte_join,"Return the highest sale price for each make and model of cars that have been sold and are no longer in inventory, ordered by the sale price from highest to lowest.","When getting a car's inventory status, always take the latest status from the inventory_snapshots table","WITH latest_inventory_status AS (SELECT car_id, is_in_inventory, ROW_NUMBER() OVER(PARTIT ...
+car_dealership,instructions_cte_join,"Return the highest sale price for each make and model of cars that have been sold and are no longer in inventory, ordered by the sale price from highest to lowest.","When getting a car's inventory status, always take the latest status from the inventory_snapshots table","WITH latest_inventory_status AS (SELECT car_id, is_in_inventory FROM inventory_snapshots) ...
```
This is a tricky one and it took me awhile to figure out what's going on here. I think the golden query seems right, and the key lies in the instruction:
> "When getting a car's inventory status, always take the latest status from the inventory_snapshots table"

ie to get the inventory status, we'll need to first get the latest row for each car in inventory_snapshot, hence the window function that partitions by car id and orders by snapshot_date (crtd_ts can be safely ignored). That said, I think this might still be ambiguous because this could mean either:
* get the latest snapshot date, and only consider cars whose is_in_inventory on that specific date. in our data, the latest snapshot date is 2023-04-15, and there is only 1 entry on 2023-04-15 (car 9), and that means only that car is considered.
* for each car, get its respective latest date. eg car 1's latest date is 2023-03-28, while car 9's latest date is 2023-04-15, and then use the is_in_inventory for that latest date. in our data this means that we have 10 cars that we can consider, out of which 8 is_in_inventory is false.

I think this is hard to clarify much further without an extensive glossary about how the tables work, and I think we might need to adapt the data, such that we get full snapshots of all cars on each snapshot date so that either of both interpretations return the same result. wdyt?

wendy-aw replied:
Ok i'm slightly confused. I'm not getting how the instructions could mean the first point. From "When getting a car's inventory status" I just thought it's pretty straightforward for it to mean an individual car, the same way your second pointer implies. Which means the row_number() is still important to the question.

rishsriv replied:
Aye! From looking at the data underneath,the same car can have multiple entries in the inventory_snapshots table. For example, if this was (say) a second hand car store that can have the same car floating in and out of the system. M

Just clarifying this in either the instructions or questions should work (without updating the underlying data). From the question + instructions, it’s unclear that this would be the case!

I'll do this in a bit :D Thanks JP and Wendy!

wongjingping replied:
Thanks for the inputs all! On a second read, I agree that the second point is the more natural interpretation and am for clarifying the question/instructions.

rishsriv replied:
Fixed in latest push!

wongjingping commented on eval/bedrock_runner.py:
````diff
@@ -38,20 +38,10 @@ def process_row(row, model_id, decimal_points):
+    generated_query = (
+        generated_query.split("```sql")[-1].split("```")[0].split(";")[0].strip() + ";"
+    )
+    print(generated_query)
````
nit: should we remove the print statement?

rishsriv replied:
Will do!

wongjingping commented on query_generators/anthropic.py:
```diff
@@ -95,18 +96,30 @@ def generate_query(
+                    shuffle,
+                )
+                pruned_metadata_str = pruned_metadata_ddl + join_str
+            elif columns_to_keep == 0:
```
thanks for adding this! sorry I missed this while updating the `columns_to_keep` earlier.

rishsriv replied:
All good!

wendy-aw commented on query_generators/anthropic.py:
```diff
@@ -95,18 +96,30 @@ def generate_query(
         with open(self.prompt_file) as file:
             model_prompt = file.read()
         question_instructions = question + " " + instructions
         if table_metadata_string == "":
```
Oooh good catch that we missed out on this logic for the anthropic runner

wongjingping reviewed (commented):
Thanks for updating, left a small comment on the advanced question.

wendy-aw reviewed (approved):
Thanks for the improvements!

rishsriv commented:
Made a final change (couldn't find this yesterday because it was a Monday)!

For this question in v1: `show me the daily total amount of payments received in the whole of last week, split by the payment_method`

Our previous golden query was:
```sql
SELECT
  payment_date,
  payment_method,
  SUM(payment_amount) AS total_amount
FROM
  payments_received
WHERE
  payment_date BETWEEN
    DATE_TRUNC('WEEK', CURRENT_DATE) - INTERVAL '1 week'
    AND DATE_TRUNC('WEEK', CURRENT_DATE)
GROUP BY
  payment_date, payment_method
ORDER BY payment_date DESC, payment_method ASC;
```

However, the question was ambiguous. Last week can mean both the last 7 days, and the last ISO week. So I have changed this to `show me the daily total amount of payments received in the whole of the last ISO week, split by the payment_method`

Additionally, the `BETWEEN` clause in inclusive. So this query would give answers for a 8 day period instead of a 7 day one. So the golden query is now updated to:

```sql
SELECT
  payment_date,
  payment_method,
  SUM(payment_amount) AS total_amount
FROM
  payments_received
WHERE
  payment_date >= DATE_TRUNC('WEEK', CURRENT_DATE) - INTERVAL '1 week'
  AND payment_date < DATE_TRUNC('WEEK', CURRENT_DATE)
GROUP BY
  payment_date, payment_method
ORDER BY payment_date DESC, payment_method ASC;
```

wendy-aw commented:
ooh good find. thanks for the change!

wongjingping reviewed (approved):
lgtm thanks for the detailed fixes and explanation as well!
